src/screenshot.c

```diff
@@ -488,37 +499,71 @@ screenshot_iface_init (XdpDbusScreenshotIface *iface)
   iface->handle_pick_color = handle_pick_color;
 }
 
+static void
+screenshot_dispose (GObject *object)
+{
+  Screenshot *screenshot = (Screenshot *) object;
+
+  g_clear_object (&screenshot->impl);
+  g_clear_object (&screenshot->access_impl);
+
+  G_OBJECT_CLASS (screenshot_parent_class)->dispose (object);
+}
+
 static void
 screenshot_init (Screenshot *screenshot)
 {
-  /* Before there was a version property, the version was hardcoded to 2, so
-   * make sure we retain that behaviour */
-  impl_version = 2;
-  xdp_dbus_screenshot_set_version (XDP_DBUS_SCREENSHOT (screenshot),
-                                   impl_version);
 }
 
 static void
 screenshot_class_init (ScreenshotClass *klass)
 {
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->dispose = screenshot_dispose;
+}
+
+static Screenshot *
+screenshot_new (XdpDbusImplScreenshot *impl,
+                XdpDbusImplAccess     *access_impl)
+{
+  Screenshot *screenshot;
+  g_autoptr(GVariant) version = NULL;
+
+  screenshot = g_object_new (screenshot_get_type (), NULL);
+  screenshot->access_impl = g_object_ref (access_impl);
+  screenshot->impl = g_object_ref (impl);
+
+  g_dbus_proxy_set_default_timeout (G_DBUS_PROXY (screenshot->impl), G_MAXINT);
+
+
+  /* Before there was a version property, the version was hardcoded to 2, so
+   * make sure we retain that behaviour */
+  version = g_dbus_proxy_get_cached_property (G_DBUS_PROXY (screenshot->impl),
+                                              "version");
+  screenshot->impl_version =
+    (version != NULL) ? g_variant_get_uint32 (version) : 2;
+
+  xdp_dbus_screenshot_set_version (XDP_DBUS_SCREENSHOT (screenshot),
+                                   screenshot->impl_version);
```

Comment:
You're missing the `MAX` check to make sure that we don't report a newer version than we actually support if the backend happens to be newer than the frontend